Threading API conversion, vol. 1

ghedo opened the pull request:
This is the first part of the conversion to the new multi-threading API. It's already quite big, so it's better to review the work so far instead of getting a bigger (and more difficult to review) PR later on.

I added a few more *_up_ref() functions which is IMO better practice instead of using the atomic add directly and there's some style cleanup mixed in (sorry about that).

FdaSilvaYY commented on crypto/bio/bio_lib.c:
```diff
@@ -94,12 +94,21 @@ int BIO_set(BIO *bio, BIO_METHOD *method)
+
+    if (method->create != NULL) {
         if (!method->create(bio)) {
-            CRYPTO_free_ex_data(CRYPTO_EX_INDEX_BIO, bio, &bio->ex_data);
```
Is there any reason to discard CRYPTO_free_ex_data call ?

ajmohan commented on crypto/bio/bio_lib.c:
```diff
@@ -94,12 +94,21 @@ int BIO_set(BIO *bio, BIO_METHOD *method)
         if (!method->create(bio)) {
-            CRYPTO_free_ex_data(CRYPTO_EX_INDEX_BIO, bio, &bio->ex_data);
-            return (0);
+            BIO_free(bio);
```
If BIO_set is invoked by BIO_new; doing BIO_free(bio) here and returning 0 will cause to do OPENSSL_free (bio) twice. First in BIO_free(bio) and second in BIO_new() [ when it sees BIO_set failed, it is invoking OPENSSL_free()]

ghedo replied:
Fixed, thanks.

ajmohan commented on crypto/bio/bio_lib.c:
```diff
@@ -94,12 +94,21 @@ int BIO_set(BIO *bio, BIO_METHOD *method)
+
+    bio->lock = CRYPTO_THREAD_lock_new();
+    if (bio->lock == NULL) {
+        BIO_free(bio);
```
This will cause double free of 'bio'.
I think doing CRYPTO_free_ex_data() here and let callee do OPENSSL_free (bio) is easier.

ghedo replied:
Fixed, thanks.

mattcaswell commented:
@ghedo, FYI I have patches for converting the init and async code, so no need to touch those.

richsalz commented:
Can you do crypto/err as a separate MR?  I'll review that one.

ajmohan commented on crypto/ec/ec_kmeth.c:
```diff
@@ -153,6 +153,14 @@ EC_KEY *EC_KEY_new_method(ENGINE *engine)
+    if (ret->lock == NULL) {
+        ECerr(EC_F_EC_KEY_NEW_METHOD, ERR_R_MALLOC_FAILURE);
+        EC_KEY_free(ret);
+        return NULL;
```
CRYPTO_free_ex_data ret->ex_data here and in other error paths.

ghedo replied:
EC_KEY_free already does that.

ajmohan replied:
Got it, thanks.

ajmohan commented on crypto/ec/ec_kmeth.c:
```diff
@@ -153,6 +153,14 @@ EC_KEY *EC_KEY_new_method(ENGINE *engine)
+    }
+
     if (ret->meth->init != NULL && ret->meth->init(ret) == 0) {
         EC_KEY_free(ret);
```
also free ret->lock

ghedo replied:
Same as above.

ajmohan replied:
Yep, thanks.

ghedo commented:
@richsalz roger roger https://github.com/openssl/openssl/pull/779

ghedo commented:
Rebased.

FdaSilvaYY commented on crypto/asn1/x_pkey.c:
```diff
@@ -89,7 +93,7 @@ void X509_PKEY_free(X509_PKEY *x)
         return;
 
-    i = CRYPTO_add(&x->references, -1, CRYPTO_LOCK_X509_PKEY);
+    CRYPTO_atomic_add(&x->references, -1, &i, x->lock);
```
ret->lock could be be NULL here.
As when X509_PKEY_new failed on CRYPTO_THREAD_lock_new() call.
Please fix.

ghedo replied:
Fixed, thanks.

ghedo replied:
I fixed a few similar issues as well.

FdaSilvaYY commented on crypto/bio/bio_lib.c:
```diff
@@ -94,12 +94,23 @@ int BIO_set(BIO *bio, BIO_METHOD *method)
+    bio->lock = CRYPTO_THREAD_lock_new();
+    if (bio->lock == NULL) {
+        CRYPTO_free_ex_data(CRYPTO_EX_INDEX_BIO, bio, &bio->ex_data);
+        CRYPTO_THREAD_lock_free(bio->lock);
```
bio->lock should be NULL here , no ?

ghedo replied:
Fixed, thanks.

ghedo commented:
@mattcaswell are you planning to convert `CRYPTO_LOCK_INIT` as well?

FdaSilvaYY commented on crypto/asn1/x_pkey.c:
```diff
@@ -75,6 +75,14 @@ X509_PKEY *X509_PKEY_new(void)
     if (ret->enc_algor == NULL || ret->enc_pkey == NULL)
         goto err;
 
+    ret->lock = CRYPTO_THREAD_lock_new();
```
IMHO, you should move this block  a few lines upper, just after ret->reference = 1;
So, the if statement will be shorter.

ghedo replied:
Done, thanks.

FdaSilvaYY commented on crypto/ec/ec_kmeth.c:
```diff
@@ -122,13 +122,20 @@ EC_KEY *EC_KEY_new_method(ENGINE *engine)
+    if (ret->lock == NULL) {
+        ECerr(EC_F_EC_KEY_NEW_METHOD, ERR_R_MALLOC_FAILURE);
+        OPENSSL_free(ret);
+        return NULL;
```
a CRYPTO_free_ex_data call is missing here.

ghedo replied:
Fixed, thanks. Following error paths were missing that as well.

FdaSilvaYY commented on crypto/ec/ec_key.c:
```diff
@@ -204,6 +204,13 @@ EC_KEY *EC_KEY_dup(EC_KEY *ec_key)
     if (ret == NULL)
         return NULL;
+
+    ret->lock = CRYPTO_THREAD_lock_new();
```
Already allocated by EC_KEY_new_method, no ?

ghedo replied:
Fixed, thanks.

FdaSilvaYY commented on crypto/x509/x509cset.c:
```diff
@@ -134,7 +134,8 @@ int X509_CRL_sort(X509_CRL *c)
 
 void X509_CRL_up_ref(X509_CRL *crl)
```
Most of <Type>_up_ref() methods are returning an int, except a few like this one.
Is it a matter for a separated PR ?

ghedo replied:
It's outside the scope of this PR IMO.

FdaSilvaYY commented on ssl/ssl_ciph.c:
```diff
@@ -547,41 +548,33 @@ static int sk_comp_cmp(const SSL_COMP *const *a, const SSL_COMP *const *b)
+            sk_SSL_COMP_push(ssl_comp_methods, comp);
+            sk_SSL_COMP_sort(ssl_comp_methods);
         }
+        CRYPTO_mem_ctrl(CRYPTO_MEM_CHECK_ENABLE);
```
The call to CRYPTO_mem_ctrl is done one accolade earlier than before, isn't it ?

ghedo replied:
Fixed, thanks.

FdaSilvaYY commented on ssl/ssl_lib.c:
```diff
@@ -2412,17 +2423,22 @@ SSL_CTX *SSL_CTX_new(const SSL_METHOD *meth)
     ret->options |= SSL_OP_NO_COMPRESSION;
 
-    return (ret);
+    ret->lock = CRYPTO_THREAD_lock_new();
```
IMHO, you should allocate it earlier, just after ret->references = 1;

ghedo replied:
Fixed, thanks.

mattcaswell commented:
@gheo, my init patch currently in review converts CRYPTO_LOCK_INIT

mattcaswell commented on crypto/asn1/tasn_utl.c:
```diff
@@ -112,15 +113,21 @@ int asn1_do_lock(ASN1_VALUE **pval, int op, const ASN1_ITEM *it)
     if (!aux || !(aux->flags & ASN1_AFLG_REFCOUNT))
         return 0;
     lck = offset2ptr(*pval, aux->ref_offset);
+    lock = offset2ptr(*pval, aux->ref_lock);
```
I don't understand this at all. Can you explain it?

mattcaswell replied:
Forget it. I think I figured it out.

mattcaswell commented on crypto/asn1/x_pubkey.c:
```diff
@@ -72,6 +72,12 @@
+        pubkey->lock = CRYPTO_THREAD_lock_new();
+        if (pubkey->lock == NULL)
+            return 0;
+    }
```
I have no idea what this is all about. Care to enlighten me?

ghedo replied:
The lock needs to be allocated somewhere since there's no `X509_PUBKEY_new` function where we can do that. I noticed that the lock didn't get freed though (in the `ASN1_OP_FREE_POST` phase), so I fixed that now.

mattcaswell commented on ssl/ssl_cert.c:
```diff
@@ -176,7 +167,14 @@ CERT *ssl_cert_new(void)
     ret->sec_ex = NULL;
-    return (ret);
+    ret->lock = CRYPTO_THREAD_lock_new();
+    if (ret == NULL) {
```
I think you mean `ret->lock` here

ghedo replied:
Fixed, thanks.

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -2785,8 +2805,9 @@ void ssl_update_cache(SSL *s, int mode)
-        && (s->session_ctx->new_session_cb != NULL)) {
-        CRYPTO_add(&s->session->references, 1, CRYPTO_LOCK_SSL_SESSION);
+        && (s->session_ctx->new_session_cb != NULL))
+    {
```
The `{` should be on the end of the previous line

ghedo replied:
Fixed.

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -749,16 +756,17 @@ SSL *SSL_new(SSL_CTX *ctx)
 {
-    CRYPTO_add(&s->references, 1, CRYPTO_LOCK_SSL);
+    int i;
+    CRYPTO_atomic_add(&s->references, 1, &i, s->lock);
```
I can't think of a good reason why we need a full lock here. SSL objects are inherently single threaded, no?

richsalz replied:
We have the up_ref already.  It is useful to prevent errors when sharing SSL objects across a worker/thread-pool model.

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -791,17 +799,17 @@ int SSL_set_session_id_context(SSL *ssl, const unsigned char *sid_ctx,
 int SSL_set_generate_session_id(SSL *ssl, GEN_SESSION_CB cb)
 {
-    CRYPTO_w_lock(CRYPTO_LOCK_SSL);
+    CRYPTO_THREAD_write_lock(ssl->lock);
```
This looks like a copy&paste error in the original code, i.e. copy&paste from the SSL_CTX_set_generate_session_id case. I can't think why the SSL object needs this too.

mattcaswell commented:
All the new `*up_ref` functions need documentation adding.

mattcaswell commented on crypto/asn1/tasn_utl.c:
```diff
@@ -112,15 +113,21 @@ int asn1_do_lock(ASN1_VALUE **pval, int op, const ASN1_ITEM *it)
 #endif
     REF_ASSERT_ISNT(ret < 0);
+    if (ret == 0)
+        CRYPTO_THREAD_lock_free(*lock);
```
Isn't this code better placed with the rest of the "freeing" code in the `ASN1_ITYPE_SEQUENCE` case of `asn1_item_embed_free`?

ghedo replied:
I don't thnk it's safe. That is, not all `ASN1_ITYPE_SEQUENCE` objects have a lock so we might end up with an invalid pointer. Though the whole ASN1 code is kind of confusing, so I may be wrong.

ghedo commented:
FYI, there seems to be a memory leak in the SSL-related commit and I'm currently investigating it.

ghedo commented:
I added documentation for `BIO_up_ref`. Regarding `X509_STORE_up_ref`, there doesn't seem to be any document for `X509_STORE` functions, so I don't know where the up_ref thing should go.

Also, I removed the SSL commit for now so the memleak in there doesn't block this. I'll open a new PR with just that commit.

mattcaswell commented:
@ghedo, if there's no existing pod file for `X509_STORE_up_ref` just create a new one to just document that one function. We've got to start somewhere!

mattcaswell commented on doc/crypto/BIO_new.pod:
```diff
@@ -38,6 +39,8 @@ BIO_new() returns a newly created BIO or NULL if the call fails.
 
 BIO_set(), BIO_free() return 1 for success and 0 for failure.
 
+BIO_up_ref() increments the reference count associated with the BIO object.
```
This should be in the DESCRIPTION section, with BIO_up_ref() additionally added to the sentence above with the list of functions returning 1 for success and 0 for failure.

ghedo replied:
Whoops. Fixed, thanks.

ghedo commented:
I added `X509_STORE_new.pod` which also documents `_new()` and `_free()`. Doesn't look like I need to make changes to the build system for that, right?

mattcaswell commented:
No, the build system should just pick it up.

mattcaswell commented:
+1, ping @richsalz

richsalz commented:
Right, no build changes.  Thanks!

richsalz commented:
Needs a rebase; let me know if you want me to do it.

ghedo commented:
@richsalz the rebase  is needed to update crypto.h, so I'll do it once #779 is merged (otherwise I'd have to do it twice) which has been rebased already.

ghedo commented:
@richsalz aaaand rebased! Turns out doing parallel PRs wasn't such a great idea after all :/ (though maybe it helped for reviewing?)

ghedo commented:
Also thanks to everyone who participated in the review: @richsalz, @mattcaswell, @FdaSilvaYY and @ajmohan (hope I didn't forget anyone, that would be embarassing)!!!

richsalz commented:
Yes, truly a team effort.  Thanks to everyone who helped on this!
